### ISSUE TITLE
Aeson 1.5-2.0 compat

### ISSUE THREAD

**Issue by tomjaguarpaw** [haskellfoundation/stability#9]
Here is a suggestion for an Aeson 1.5-2.0 compatibilty story. The technique is simple: take Aeson 2.0 and change the version 2 abstract types to type synonyms as below (plus chasing up the ensuing, minor, type errors).

This backports the 2.0 API to 1.5, and ought to allow Aeson users to write code that works at the same time against version 1.5 and version 2.0, providing a smooth upgrade path and avoiding the need for a "big bang" switch of the entire Aeson ecosystem at once.

```diff
-newtype Key = Key { unKey :: Text }
+type Key = Text

-newtype KeyMap v = KeyMap { unKeyMap :: Map Key v }
+type KeyMap v = Map Key v
```

I don't actually use Aeson myself, so I would welcome the input of Aeson users and maintainers (esp. @bergmark) regarding

1. whether this is actually useful, and if so
2. whether I've got everything correct (the tests pass)

* See [my branch and the full diff](https://github.com/haskell/aeson/compare/v2.0.0.0...tomjaguarpaw:1.5-2.0-compat?expand=1)

**Comment by NorfairKing** [comment]
+1 for a CPP-less migration option!

**Comment by bergmark** [comment]
I think this is a good addition. 

One upside is that it doesn't require usage of a compat package. 

The downside is that every major aeson series we wish to support for this migration needs to be amended with this patch. Bounds on dependents would need to be adjusted to e.g. `^>= 1.5.7 || ^>= 2.0`. I think that would be fine, as aeson 1.5.x has been out for almost two years and has pretty good backwards compat for its dependencies. And if someone needs more backwards compat on aeson they are free to not use this.

The compat package approach (we could always do both) would allow us to support more versions of aeson (bounds could be simplified to e.g. `>= 1 && < 1.6 || ^>= 2.0`) with only one release of that package.

FWIW I think that filing this directly against aeson would be better. I'm one of three maintainers but I'm currently the least active one so I think it's better to discuss it over there.

Just noting a detail in case this turns into a full PR: The change should probably be made against the latest 1.5.x commit.

**Comment by tomjaguarpaw** [comment]
OK, migrated to Aeson at https://github.com/haskell/aeson/pull/932

**Comment by cdornan** [comment]
Reopening as there seems to be some difference of opinion as to whether this belongs in Aeson.

**Comment by tomjaguarpaw** [comment]
I think I may not have explained very well what my suggestion actually is. It seems that it was interpreted to be a change to 2.0 to revert making `Key` and `KeyMap` abstract.

**Comment by cdornan** [comment]
Indeed! Context is key -- we all understood what you intended but Oleg will have been missing that context.

Let's discuss later.
